File: programs-basic-java/src/main/java/com/local/numbertypes/SpyNumber.java
package com.local.numbertypes;

import lombok.extern.slf4j.Slf4j;

@Slf4j
public class SpyNumber implements INumberTypeOperation {
	// A number whose sum of the digits of is equal to the product of its digits
	// 132
	// The sum of digits = 1 + 3 + 2 = 6
	// The product of the digits = 1 * 3 * 2 = 6

	long factorial(long number) {
		return (number == 1l || number == 0l) ? 1 : number * factorial(number - 1);
	}

	@Override
	public boolean execute(long input, long order) {
		var sum = 0l;
		var product = 1l;
		var number = input;
		while (number != 0) {
			var lastDigit = number % 10;
			sum = sum + lastDigit;
			product = product * lastDigit;
			number = number / 10;
		}
		var status = (sum == product);
		log(status, input);
		return status;
	}

	private void log(boolean status, long input) {
		if (status) {
			log.debug("An Spy Number {}", input);
		} else {
			log.debug("Not an Spy Number {}", input);
		}
	}

}

File: programs-basic-kotlin/src/main/kotlin/com/local/numbertypes/SpyNumber.kt
package com.local.numbertypes

import mu.KotlinLogging
private val logger = KotlinLogging.logger {}
class SpyNumber : INumberTypeOperation {
	// A number whose sum of the digits of is equal to the product of its digits
	// 132
	// The sum of digits = 1 + 3 + 2 = 6
	// The product of the digits = 1 * 3 * 2 = 6
	fun factorial(number: Long): Long {
		return if (number == 1L || number == 0L) 1 else number * factorial(number - 1)
	}

	override fun execute(input: Long, order: Long): Boolean {
		var sum = 0L
		var product = 1L
		var number = input
		while (number != 0L) {
			val lastDigit = number % 10
			sum += lastDigit
			product *= lastDigit
			number /= 10
		}
		val status = sum == product
		log(status, input)
		return status
	}

	private fun log(status: Boolean, input: Long) {
		if (status) {
			logger.debug("An Spy Number {}", input)
		} else {
			logger.debug("Not an Spy Number {}", input)
		}
	}
}